File: Mancunian and K-Ordered LCS.cpp
//https://www.hackerearth.com/problem/algorithm/mancunian-and-k-ordered-lcs-e6a4b8c6/submissions/

#include<iostream>
#include<algorithm>
using namespace std;
long int lcs(long int n,long int m,vector <int> a,vector <int> b)
{
    vector<vector<int>> arr (n + 1, vector<int>(m + 1));
    long int i,j;
    for(i=0;i<n;i++)
    {
        for(j=0;j<m;j++)
        {
            if(a[i]==b[j])
            {
                arr[i+1][j+1]=arr[i][j]+1;
            }
            else
            {
                arr[i+1][j+1]=max(arr[i+1][j],arr[i][j+1]);
            }
        }
    }
    return arr[n][m];
}
int main()
{
    long int n,m,k;
    cin>>n>>m>>k;
	vector <int> a(n);
	vector <int> b(m);
    for(long int i=0;i<n;i++)
    {
        cin>>a[i];
    }
    for(long int i=0;i<m;i++)
    {
        cin>>b[i];
    }
    cout<<lcs(n,m,a,b)+k;
    return 0;
}

File: Mancunian and K-Ordered LCS.py
#https://www.hackerearth.com/problem/algorithm/mancunian-and-k-ordered-lcs-e6a4b8c6/submissions/

def lcs(n,m,a,b):
    arr=[[0]*(m+1) for _ in range(n+1)]
    for i in range(n):
        for j in range(m):
            if(a[i]==b[j]):
                arr[i+1][j+1]=arr[i][j]+1
            else:
                arr[i+1][j+1]=max(arr[i+1][j],arr[i][j+1])
    return arr[-1][-1]
if __name__=="__main__":
    n,m,k=map(int,input().split())
    a=list(map(int,input().split()))
    b=list(map(int,input().split()))
    x=m-len(b)
    for _ in range(x):
        c=int(input())
        b.append(c)
    print(lcs(n,m,a,b)+k)